src/zenml/model/model_version.py

```diff
@@ -0,0 +1,521 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""ModelVersion user facing interface to pass into pipeline or step."""
+
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    List,
+    Optional,
+    Union,
+)
+from uuid import UUID
+
+from pydantic import BaseModel, PrivateAttr, root_validator
+
+from zenml.enums import ModelStages
+from zenml.exceptions import EntityExistsError
+from zenml.logger import get_logger
+
+if TYPE_CHECKING:
+    from zenml.models import (
+        ArtifactResponseModel,
+        ModelResponseModel,
+        ModelVersionResponseModel,
+        PipelineRunResponseModel,
+    )
+
+logger = get_logger(__name__)
+
+
+class ModelVersion(BaseModel):
+    """ModelVersion class to pass into pipeline or step to set it into a model context.
+
+    name: The name of the model.
+    license: The license under which the model is created.
+    description: The description of the model.
+    audience: The target audience of the model.
+    use_cases: The use cases of the model.
+    limitations: The known limitations of the model.
+    trade_offs: The tradeoffs of the model.
+    ethics: The ethical implications of the model.
+    tags: Tags associated with the model.
+    version: The model version name, number or stage is optional and points model context
+        to a specific version/stage. If skipped new model version will be created.
+    save_models_to_registry: Whether to save all ModelArtifacts to Model Registry,
+        if available in active stack.
+    """
+
+    name: str
+    license: Optional[str]
+    description: Optional[str]
+    audience: Optional[str]
+    use_cases: Optional[str]
+    limitations: Optional[str]
+    trade_offs: Optional[str]
+    ethics: Optional[str]
+    tags: Optional[List[str]]
+    version: Optional[Union[ModelStages, int, str]]
+    save_models_to_registry: bool = True
+
+    suppress_class_validation_warnings: bool = False
+    was_created_in_this_run: bool = False
+
+    _model_id: UUID = PrivateAttr(None)
+    _id: UUID = PrivateAttr(None)
+    _number: int = PrivateAttr(None)
+
+    #########################
+    #    Public methods     #
+    #########################
+    @property
+    def all_model_versions(self) -> List["ModelVersion"]:
```

Comment:
Doesn't `ModelVersion` already refer to a specific version of a model? In that case, what does `model_version.all_model_versions` do? Logically that should just return `[self]`. If you want all versions of the same model, then a user-facing syntax like `model_version.model.versions` would make more sense IMO

Reply:
Yes, it refers to a specific model version indeed. Maybe this one is not needed and one should `Client().list_model_versions` instead.

Reply:
Removed it completely - let users call client